### ISSUE TITLE
Web Component support

### ISSUE THREAD

**Issue by uosis** [raquo/Laminar#59]
I was wondering if there is any interest in supporting Web Components in Laminar? I am not sure if there is anything that actually needs to change in the core project, perhaps just having an example of a proper way to add and use Web Components. This would hopefully alleviate the complaint that Laminar has no ready to use components.

I got a proof of concept working as follows:

1. In the host page's head, add:

```html
  <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500" rel="stylesheet">
  <link href="https://fonts.googleapis.com/css?family=Material+Icons&display=block" rel="stylesheet">
  <script type="module" src="https://unpkg.com/@material/mwc-button?module"></script>
```

2. Now you have `mwc-button` html tag that behaves just like any other tag. Use it in Laminar (this is just a random button added in TodoMVC example):

```scala
L.htmlTag("mwc-button", void = false)(
        new ReactiveHtmlAttr[String]("label", StringAsIsCodec) <-- itemsVar.signal.map(i => s"${i.count(!_.completed)} items"))
```

In summary, the process to add a Web Component is:

1. Import component's javascript
2. Define Laminar tag / attributes based on [component's documentation](https://github.com/material-components/material-components-web-components/tree/master/packages/button)

What is the best way to do those two things? In particular:

1. Referencing an npm module from unpkg like this, while functional, is very inefficient - those modules should somehow be bundled with the rest of js. How can this be done?
2. What is the proper way to define custom tags? I didn't see any docs for this. Also once defined, what's the best way to make them available to the app?

Thanks for making Laminar!

**Comment by raquo** [comment]
Hey, I don't know much about Web Components other than the general idea, but you're right, they're useful, and this is something that should be documented.

I'm up for writing a documentation section about this if you can help me figure out how to use this button Web Component.

I've just pushed a `web-components` branch to the laminar-examples repo, check it out.

[MwcButton.scala](https://github.com/raquo/laminar-examples/blob/web-components/src/main/scala/webcomponents/MwcButton.scala) is our wrapper for the @material/mwc-button web component.

* It shows how to create custom elements – you basically got this already
* I'm using standard scala.js @JSImport to load the javascript. It works but there might be a way to avoid `val _`, not sure.
* I'm still loading CSS via the html file though because if we loaded it asynchronously, only when the component is instantiated, there would be a delay.

[WebComponentsPage.scala](https://github.com/raquo/laminar-examples/blob/web-components/src/main/scala/webcomponents/WebComponentsPage.scala) contains the usage of this component from Laminar:

```scala
MwcButton(
  _.label <-- actionVar.signal,
  _.icon := "code"
)
```

This API is just something I came up with without thinking too much. It works but better solutions are possible depending on preferences / requirements.

This setup works, however I have no idea how to register a click event listener on this button. If you know how to do this in Javascript, I can translate it into Laminar pretty easily unless we run into https://github.com/scala-js/scala-js-dom/issues/191 or https://github.com/scala-js/scala-js-dom/issues/351 (but worst case we can use js.Dynamic).

Also, I'm not sure if there is anything I should be doing to manage the lifecycle of the web component. Do I need to call `destroy` on it when it's unmounted or something? I don't know, and once again, if I knew what to do in JS I could solve it in Laminar easily. Or even if I could see a simple example of how this web component is used from some other UI library, that might help too.

**Comment by raquo** [comment]
Cool, thanks for your help too, I merged the examples web-components branch into master.

> I will also create material component library using this pattern when I get a chance, in a separate repo.

That would be so awesome, people have been asking for this kind of thing forever! It doesn't even need to be comprehensive from the start, I think you'll find contributors for such a project pretty easily if you just start with something.

Cheers

**Comment by mathieuleclaire** [comment]
Hi,
is there any effort to start a repository with Web Component wrapping so that they can be directly be used in a web ui construction ? Or is there only Button, Slider, ProgressBar wrappers so far ?

Thanks for this amazing work !

**Comment by raquo** [comment]
Hi, there is a WIP generator for many material UI web component interfaces
for Laminar, see here: https://github.com/uosis/laminar-web-components

If I understand correctly it can already be used as a decent starting point
but it would be great to implement the missing functionality of the
generator too.

--Nikita

**Comment by mathieuleclaire** [comment]
Excellent, thanks for sharing. I will try to work with this.
Peharp's you could mention it on your front web page and/or rely on it in your laminar example ?

**Comment by raquo** [comment]
The project is not done yet so I can't rely on it. The repo is mentioned in
a few places including the video if I recall correctly.

--Nikita
